src/session/content/message_row/document.rs

```diff
@@ -0,0 +1,357 @@
+                s.finish() as usize
+            }) % classes.len()];
+            imp.sender_label.add_css_class(color_class);
+
+            *sender_color_class = Some(color_class.into());
+
+            imp.sender_label.set_visible(true);
+        } else {
+            imp.sender_label.set_visible(false);
+        }
+
+        let handler_id =
+            message.connect_content_notify(clone!(@weak self as obj => move |message, _| {
+                obj.update_document(message);
+            }));
+        imp.handler_id.replace(Some(handler_id));
+        self.update_document(&message);
+
+        imp.message.replace(Some(message));
+        self.notify("message");
+    }
+}
+
+#[derive(PartialEq)]
+enum FileStatus {
+    Downloading(f64),
+    Uploading(f64),
+    CanBeDownloaded,
+    Downloaded,
+}
+use FileStatus::*;
+
+impl From<&File> for FileStatus {
+    fn from(file: &File) -> Self {
+        let local = &file.local;
+        let remote = &file.remote;
+
+        let size = file.size.max(file.expected_size) as u64;
+
+        if local.is_downloading_active {
+            let progress = local.downloaded_size as f64 / size as f64;
+            Downloading(progress)
+        } else if remote.is_uploading_active {
+            let progress = remote.uploaded_size as f64 / size as f64;
+            Uploading(progress)
+        } else if local.is_downloading_completed {
+            Downloaded
+        } else if local.can_be_downloaded {
+            CanBeDownloaded
+        } else {
+            dbg!(file);
+            unimplemented!("unknown file status");
+        }
+    }
+}
+
+impl MessageDocument {
+    fn update_document(&self, message: &Message) {
+        if let MessageContent::MessageDocument(data) = message.content().0 {
+            let imp = self.imp();
+
+            let message_text = parse_formatted_text(data.caption);
+            imp.content_label.set_visible(!message_text.is_empty());
+            imp.content_label.set_label(message_text);
+
+            imp.file_name_label.set_label(&data.document.file_name);
+
+            let session = message.chat().session();
+
+            self.update_status(data.document.document, session);
+        }
+    }
+
+    fn update_status(&self, file: File, session: Session) -> FileStatus {
+        let status = FileStatus::from(&file);
+
+        let size = file.size.max(file.expected_size) as u64;
+
+        self.update_size_label(&status, size);
+        self.update_button(file, session, &status);
+
+        status
+    }
+
+    fn block_click(&self, block: bool) {
+        let imp = self.imp();
+        let click = &*imp.click;
+        let document_box = &*imp.document_box;
+
+        let handler_id = imp.click_handler_id.get().unwrap();
+        if block {
+            click.block_signal(handler_id);
+            document_box.remove_css_class("activatable");
+        } else {
+            click.unblock_signal(handler_id);
+            document_box.add_css_class("activatable");
```

Comment:
Mh, I guess we can just set the button to be insensitive via the "sensitive" property. I guess this will make the button grayed out in some way, probably by adding opacity or something, but we can override that style in css later.

Reply:
The button should stay clickable the whole time. It's just the message box that is deactivated here. What we currently do is
1. You can click the whole message or the button to download
2. You can only click the button to cancel the download
3. You can again click the whole message to view content

Reply:
Oh ok, but why cancelling the download only via the button? I'd expect to react the same for everything.

Reply:
@YuraIz and I discussed it in some other thread because I was thinking like you. But it seems Yuralz is correct 

https://github.com/melix99/telegrand/pull/372#discussion_r967819701